File: NoTomorrow/Services/AttendanceSync.swift
import Foundation

/// Hands my attendance writes to the backend, paired or not: the server's reminder and 21:00 "did you skip?" jobs
/// read my status for the day, and a partner sees it (a finished workout is `attended`, not a stale `confirmed`).
/// Local SwiftData stays the source of truth; this is fire and forget, with `AttendanceOutbox` for retries.
@MainActor
enum AttendanceSync {
    typealias Reporter = @MainActor (_ day: Date, _ status: AttendanceStatus) -> Void

    /// Where reports go. The default sends through `BroShared.service`; tests swap in a spy. Unit tests never
    /// reach the network: the default does nothing inside an XCTest run (the host app may be signed in).
    static var reporter: Reporter = { day, status in
        guard !isRunningTests else { return }
        Task { await BroShared.service.reportAttendance(day: day, status: status) }
    }

    static func report(day: Date, status: AttendanceStatus) {
        reporter(day, status)
    }

    /// Reports the local writes a workout edit or delete made (see `AttendanceService.wireStatus`).
    static func report(_ changes: [AttendanceService.WorkoutDayChange], isGymDay: (Date) -> Bool) {
        for change in changes {
            guard let wire = AttendanceService.wireStatus(for: change, isGymDay: isGymDay) else { continue }
            reporter(wire.day, wire.status)
        }
    }

    private static let isRunningTests = ProcessInfo.processInfo.environment["XCTestConfigurationFilePath"] != nil
}

/// My attendance writes the backend has not confirmed yet, one per day and account (the latest status wins), in
/// UserDefaults under `nt.attendance.outbox`, the same JSON array Android keeps:
/// `[{"day": "yyyy-MM-dd", "status": "cancelled", "reason": "sick", "note": "…", "makeUpDay": "yyyy-MM-dd",
/// "backend": "remote", "account": "<user id>"}]`. Every field after `status` is optional.
/// `BroService` puts a write here before sending it and removes it once the server took it, so a write made offline
/// (or cut off by the app being killed) goes out with a later refresh. Each entry carries the backend and account it
/// was made on (`Target`), and only the matching ones are sent: a write queued before a sign-out or a flip of the demo
/// switch waits for its own account instead of reaching another one.
struct AttendanceOutbox {
    /// The backend (the demo one or the real one) and the account a write was made on. The demo backend has no account.
    struct Target: Hashable {
        static let demoBackend = "demo"
        static let remoteBackend = "remote"

        var backend: String
        var account: String?

        static let demo = Target(backend: demoBackend, account: nil)

        static func remote(_ userId: String) -> Target {
            Target(backend: remoteBackend, account: userId)
        }
    }

    /// One queued write: my `status` for `day`; a cancellation also carries its reason, note and make-up day.
    struct Entry: Codable, Hashable {
        /// "yyyy-MM-dd" in the local calendar, the form the server's `/attendance/:day` takes.
        var day: String
        var status: AttendanceStatus
        var reason: String?
        var note: String?
        /// "yyyy-MM-dd", like `day`.
        var makeUpDay: String?
        /// `Target.backend`; nil only for an entry queued before entries were tagged: whoever is signed in sends it.
        var backend: String?
        var account: String?

        init(day: String, status: AttendanceStatus, reason: String? = nil, note: String? = nil,
             makeUpDay: String? = nil, target: Target? = nil) {
            self.day = day
            self.status = status
            self.reason = reason
            self.note = note
            self.makeUpDay = makeUpDay
            backend = target?.backend
            account = target?.account
        }

        var target: Target? { backend.map { Target(backend: $0, account: account) } }

        func belongs(to target: Target) -> Bool {
            self.target == nil || self.target == target
        }
    }

    static let key = "nt.attendance.outbox"
    let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// Oldest day first. Unreadable entries are dropped; the next write rewrites the list without them.
    var entries: [Entry] {
        guard let data = defaults.data(forKey: Self.key),
              let list = try? JSONDecoder().decode([Lossy].self, from: data) else { return [] }
        return list.compactMap(\.entry).sorted { $0.day < $1.day }
    }

    /// The entries `target` sends, oldest day first.
    func entries(for target: Target) -> [Entry] {
        entries.filter { $0.belongs(to: target) }
    }

    /// Queues `entry`, replacing anything queued for that day on the same backend and account.
    func put(_ entry: Entry) {
        var list = entries.filter { !($0.day == entry.day && ($0.target == entry.target || $0.target == nil)) }
        list.append(entry)
        write(list)
    }

    /// Queues `status` for `day` on `target`.
    func put(day: Date, status: AttendanceStatus, target: Target? = nil) {
        put(Entry(day: WireDay.string(day), status: status, target: target))
    }

    /// Drops `entry` once sent, unless a newer status for the same day replaced it in the meantime.
    func remove(_ entry: Entry) {
        let list = entries
        guard list.contains(entry) else { return }
        write(list.filter { $0 != entry })
    }

    /// Forgets everything queued for `target` (its account is gone), or everything when it is nil.
    func clear(target: Target? = nil) {
        guard let target else {
            defaults.removeObject(forKey: Self.key)
            return
        }
        write(entries.filter { $0.target != target })
    }

    /// A reply that retrying cannot fix (the server rejected the write itself), as opposed to no connection,
    /// a lapsed session, rate limiting or server trouble, which keep the write queued.
    static func isRejected(_ error: Error) -> Bool {
        guard case .http(let status, _, _) = BackendError.wrap(error) else { return false }
        return (400..<500).contains(status) && ![401, 408, 429].contains(status)
    }

    /// One element of the stored array; one that does not decode (an unknown status) is skipped, not fatal.
    private struct Lossy: Decodable {
        var entry: Entry?

        init(from decoder: Decoder) throws {
            entry = try? Entry(from: decoder)
        }
    }

    private func write(_ list: [Entry]) {
        if list.isEmpty {
            defaults.removeObject(forKey: Self.key)
        } else if let data = try? JSONEncoder().encode(list.sorted { $0.day < $1.day }) {
            defaults.set(data, forKey: Self.key)
        }
    }
}

File: android/app/src/main/java/app/notomorrow/service/AttendanceSync.kt
package app.notomorrow.service

import app.notomorrow.data.prefs.AppPrefs
import app.notomorrow.model.AttendanceStatus
import app.notomorrow.net.BackendError
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import java.time.LocalDate
import java.time.format.DateTimeParseException

/**
 * Hands my attendance writes to the backend, paired or not — `AttendanceSync.swift`. The server's
 * reminder and 21:00 "did you skip?" jobs read my status for the day, and a partner sees it (a
 * finished workout is `attended`, not a stale `confirmed`). Room stays the source of truth; this is
 * fire and forget, with [AttendanceOutbox] for retries.
 *
 * The app's reporter (`AppContainer.attendanceReporter`) launches [BroService.reportAttendance] on
 * the process scope, so a report outlives the screen that made it. Tests pass [None] or a spy, so
 * they never reach a backend.
 */
fun interface AttendanceReporter {

    fun report(day: LocalDate, status: AttendanceStatus)

    /** Reports the local writes a workout edit or delete made (see [AttendanceService.wireStatus]). */
    fun report(changes: List<AttendanceService.WorkoutDayChange>, isGymDay: (LocalDate) -> Boolean) {
        for (change in changes) {
            val (day, status) = AttendanceService.wireStatus(change, isGymDay) ?: continue
            report(day, status)
        }
    }

    companion object {
        /** Reports nothing: previews, tests, and the onboarding flow's own services. */
        val None = AttendanceReporter { _, _ -> }
    }
}

/**
 * My attendance writes the backend has not confirmed yet, one per day and account (the latest
 * status wins) — `AttendanceOutbox` in `AttendanceSync.swift`, kept as a JSON array under
 * `nt.attendance.outbox`:
 * `[{"day": "yyyy-MM-dd", "status": "cancelled", "reason": "sick", "note": "…", "makeUpDay": "yyyy-MM-dd",
 * "backend": "remote", "account": "<user id>"}]`. Every field after `status` is optional.
 *
 * [BroService] puts a write here before sending it and removes it once the server took it, so a
 * write made offline (or cut off by the process dying) goes out with a later refresh. Each entry
 * carries the backend and account it was made on ([Target]), and only the matching ones are sent:
 * a write queued before a sign-out or a flip of the demo switch waits for its own account instead
 * of reaching another one.
 */
class AttendanceOutbox(private val store: Store = InMemoryStore()) {

    /** Where the JSON lives. [update] must be atomic (one DataStore edit in the app). */
    interface Store {
        suspend fun read(): String?

        /** Replaces the stored JSON with `transform(current)`; `null` removes it. */
        suspend fun update(transform: (String?) -> String?)
    }

    /** Tests, and a [BroService] that never reports (the onboarding flow's own). */
    class InMemoryStore(initial: String? = null) : Store {
        private val lock = Mutex()

        var value: String? = initial
            private set

        override suspend fun read(): String? = lock.withLock { value }

        override suspend fun update(transform: (String?) -> String?) = lock.withLock { value = transform(value) }
    }

    /**
     * The backend (the demo one or the real one) and the account a write was made on. The demo
     * backend has no account.
     */
    data class Target(val backend: String, val account: String?) {
        companion object {
            const val DEMO = "demo"
            const val REMOTE = "remote"

            val Demo = Target(DEMO, null)

            fun remote(userId: String) = Target(REMOTE, userId)
        }
    }

    /**
     * One queued write: my [status] for [day]; a cancellation also carries its [reason], [note]
     * and [makeUpDay]. [target] is `null` only for an entry queued before entries were tagged:
     * whoever is signed in sends it.
     */
    data class Entry(
        val day: LocalDate,
        val status: AttendanceStatus,
        val reason: String? = null,
        val note: String? = null,
        val makeUpDay: LocalDate? = null,
        val target: Target? = null,
    ) {
        fun belongsTo(target: Target): Boolean = this.target == null || this.target == target
    }

    /** Oldest day first. */
    suspend fun entries(): List<Entry> = decode(store.read())

    /** The entries [target] sends, oldest day first. */
    suspend fun entries(target: Target): List<Entry> = entries().filter { it.belongsTo(target) }

    /** Queues [entry], replacing anything queued for that day on the same backend and account. */
    suspend fun put(entry: Entry) {
        store.update { json ->
            val kept = decode(json).filterNot { it.day == entry.day && (it.target == entry.target || it.target == null) }
            encode(kept + entry)
        }
    }

    /** Queues [status] for [day] on [target]. */
    suspend fun put(day: LocalDate, status: AttendanceStatus, target: Target? = null) =
        put(Entry(day, status, target = target))

    /** Drops [entry] once sent, unless a newer status for the same day replaced it in the meantime. */
    suspend fun remove(entry: Entry) {
        store.update { json ->
            val list = decode(json)
            if (entry in list) encode(list - entry) else json
        }
    }

    /** Forgets everything queued for [target], or everything when it is `null`. */
    suspend fun clear(target: Target? = null) {
        if (target == null) {
            store.update { null }
        } else {
            store.update { json -> encode(decode(json).filterNot { it.target == target }) }
        }
    }

    @Serializable
    private data class Wire(
        val day: String,
        val status: String,
        val reason: String? = null,
        val note: String? = null,
        val makeUpDay: String? = null,
        val backend: String? = null,
        val account: String? = null,
    )

    companion object {

        private val json = Json { ignoreUnknownKeys = true }
        private val serializer = ListSerializer(Wire.serializer())

        /** `nt.attendance.outbox` in DataStore. */
        fun prefsStore(prefs: AppPrefs): Store = object : Store {
            override suspend fun read(): String? = prefs.attendanceOutboxOnce()
            override suspend fun update(transform: (String?) -> String?) = prefs.updateAttendanceOutbox(transform)
        }

        /**
         * A reply that retrying cannot fix (the server rejected the write itself), as opposed to no
         * connection, a lapsed session, rate limiting or server trouble, which keep the write queued.
         */
        fun isRejected(error: Throwable): Boolean {
            val wrapped = BackendError.wrap(error)
            val status = (wrapped as? BackendError.Http)?.status ?: return false
            return status in 400..499 && status !in RETRYABLE_4XX
        }

        private val RETRYABLE_4XX = setOf(401, 408, 429)

        /** Unreadable or unknown entries are dropped; the next write rewrites the list without them. */
        internal fun decode(raw: String?): List<Entry> {
            if (raw.isNullOrEmpty()) return emptyList()
            val wire = runCatching { json.decodeFromString(serializer, raw) }.getOrElse { return emptyList() }
            return wire.mapNotNull { item ->
                val day = parseDay(item.day) ?: return@mapNotNull null
                val status = AttendanceStatus.from(item.status) ?: return@mapNotNull null
                Entry(
                    day = day,
                    status = status,
                    reason = item.reason,
                    note = item.note,
                    makeUpDay = item.makeUpDay?.let(::parseDay),
                    target = item.backend?.let { Target(it, item.account) },
                )
            }.sortedBy { it.day }
        }

        /** Sorted by day; `null` for an empty list, so the key goes away. Empty fields are left out. */
        internal fun encode(list: List<Entry>): String? {
            if (list.isEmpty()) return null
            return json.encodeToString(
                serializer,
                list.sortedBy { it.day }.map {
                    Wire(
                        day = it.day.toString(),
                        status = it.status.raw,
                        reason = it.reason,
                        note = it.note,
                        makeUpDay = it.makeUpDay?.toString(),
                        backend = it.target?.backend,
                        account = it.target?.account,
                    )
                },
            )
        }

        private fun parseDay(raw: String): LocalDate? = try {
            LocalDate.parse(raw)
        } catch (_: DateTimeParseException) {
            null
        }
    }
}
